File: Packages/EditorKit/Sources/EditorKit/ArrowGeometry.swift
import CoreGraphics

public enum ArrowGeometry {
    /// The two arrowhead wing points for an arrow from `start` to `end`.
    public static func headWings(start: CGPoint, end: CGPoint,
                                 length: CGFloat, halfAngleDegrees: CGFloat)
        -> (left: CGPoint, right: CGPoint) {
        let dx = end.x - start.x, dy = end.y - start.y
        let angle = atan2(dy, dx)
        let half = halfAngleDegrees * .pi / 180
        let a1 = angle + .pi - half
        let a2 = angle + .pi + half
        let left = CGPoint(x: end.x + length * cos(a1), y: end.y + length * sin(a1))
        let right = CGPoint(x: end.x + length * cos(a2), y: end.y + length * sin(a2))
        return (left, right)
    }

    /// Where the shaft should stop so it ends at the arrowhead's base instead of
    /// the tip — otherwise a thick round line cap pokes out past the tip. The
    /// head's depth along the arrow axis is `headLength·cos(halfAngle)`; the
    /// shaft is clamped so it never points backwards for a very short arrow.
    public static func shaftEnd(start: CGPoint, end: CGPoint,
                                headLength: CGFloat, halfAngleDegrees: CGFloat) -> CGPoint {
        let dx = end.x - start.x, dy = end.y - start.y
        let len = (dx * dx + dy * dy).squareRoot()
        guard len > 0 else { return start }
        let backset = headLength * cos(halfAngleDegrees * .pi / 180)
        let shaftLen = max(0, len - backset)
        let t = shaftLen / len
        return CGPoint(x: start.x + dx * t, y: start.y + dy * t)
    }
}

File: windows/src/BetterScreenshot.Editor/ArrowGeometry.cs
using BetterScreenshot.Core;

namespace BetterScreenshot.Editor;

/// <summary>
/// Pure arrow geometry: the two wing points of the arrowhead triangle, and where the shaft should end so its
/// round cap does not bleed past the arrowhead. Head half-angle defaults to 28°.
/// </summary>
public static class ArrowGeometry
{
    /// <summary>
    /// The two wing points of the arrowhead at <paramref name="end"/>, each <paramref name="length"/> from the tip,
    /// splayed ±<paramref name="halfAngleDegrees"/> from the shaft direction.
    /// </summary>
    public static (PxPoint Left, PxPoint Right) HeadWings(PxPoint start, PxPoint end, double length, double halfAngleDegrees = 28)
    {
        double dx = end.X - start.X, dy = end.Y - start.Y;
        double len = Math.Sqrt(dx * dx + dy * dy);
        if (len < 1e-9) return (end, end);

        double ux = dx / len, uy = dy / len; // forward unit vector
        double ha = halfAngleDegrees * Math.PI / 180.0;

        var left = end.Offset(length * RotX(ux, uy, Math.PI - ha), length * RotY(ux, uy, Math.PI - ha));
        var right = end.Offset(length * RotX(ux, uy, Math.PI + ha), length * RotY(ux, uy, Math.PI + ha));
        return (left, right);
    }

    /// <summary>
    /// Where the shaft should stop: back from the tip by the arrowhead depth (headLength·cos(halfAngle)). For a
    /// short arrow whose head would be longer than the shaft, this clamps to <paramref name="start"/>.
    /// </summary>
    public static PxPoint ShaftEnd(PxPoint start, PxPoint end, double headLength, double halfAngleDegrees = 28)
    {
        double dx = end.X - start.X, dy = end.Y - start.Y;
        double len = Math.Sqrt(dx * dx + dy * dy);
        if (len < 1e-9) return start;

        double headDepth = headLength * Math.Cos(halfAngleDegrees * Math.PI / 180.0);
        double shaftLen = len - headDepth;
        if (shaftLen <= 0) return start;

        double ux = dx / len, uy = dy / len;
        return new PxPoint(start.X + ux * shaftLen, start.Y + uy * shaftLen);
    }

    private static double RotX(double ux, double uy, double theta) => ux * Math.Cos(theta) - uy * Math.Sin(theta);
    private static double RotY(double ux, double uy, double theta) => ux * Math.Sin(theta) + uy * Math.Cos(theta);
}
